File: src/test/java/com/everyonewaiter/domain/auth/AuthCodeTest.java
package com.everyonewaiter.domain.auth;

import static com.everyonewaiter.domain.auth.AuthFixture.createAuthCode;
import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatCode;
import static org.assertj.core.api.AssertionsForClassTypes.assertThatThrownBy;

import java.time.Duration;
import org.junit.jupiter.api.Test;

class AuthCodeTest {

  @Test
  void verify() {
    AuthCode authCode = createAuthCode();

    assertThatCode(() -> authCode.verify(123456)).doesNotThrowAnyException();
  }

  @Test
  void verifyFail() {
    AuthCode authCode = createAuthCode();

    assertThatThrownBy(() -> authCode.verify(0))
        .isInstanceOf(ExpiredVerificationCodeException.class);
    assertThatThrownBy(() -> authCode.verify(999999))
        .isInstanceOf(UnmatchedVerificationCodeException.class);
  }

  @Test
  void key() {
    AuthCode authCode = createAuthCode();

    assertThat(authCode.key()).isEqualTo("auth:code:01012345678");
  }

  @Test
  void value() {
    AuthCode authCode = createAuthCode();

    assertThat(authCode.value()).isEqualTo(123456);
  }

  @Test
  void expiration() {
    AuthCode authCode = createAuthCode();

    assertThat(authCode.expiration()).isEqualTo(Duration.ofMinutes(5));
  }

}

File: src/test/java/com/everyonewaiter/domain/auth/AuthCodeTest.kt
package com.everyonewaiter.domain.auth

import org.assertj.core.api.Assertions.*
import org.junit.jupiter.api.Test
import java.time.Duration

class AuthCodeTest {

  @Test
  fun `인증 코드 검증`() {
    val code = createAuthCode()

    assertThatCode { code.verify(code.code()) }.doesNotThrowAnyException()
  }

  @Test
  fun `인증 코드가 만료된 경우 검증 실패`() {
    val code = createAuthCode()

    assertThatThrownBy { code.verify(0) }.isInstanceOf(ExpiredVerificationCodeException::class.java)
  }

  @Test
  fun `인증 코드가 일치하지 않는 경우 검증 실패`() {
    val code = createAuthCode()

    assertThatThrownBy { code.verify(999999) }.isInstanceOf(UnmatchedVerificationCodeException::class.java)
  }

  @Test
  fun `인증 코드 저장용 키`() {
    val code = createAuthCode()

    assertThat(code.key()).isEqualTo("auth:code:${code.phoneNumber.value}")
  }

  @Test
  fun `인증 코드 저장용 값`() {
    val code = createAuthCode()

    assertThat(code.value()).isEqualTo(code.code())
  }

  @Test
  fun `인증 코드의 유효기간은 5분`() {
    val code = createAuthCode()

    assertThat(code.expiration).isEqualTo(Duration.ofMinutes(5))
  }
}
